
#This code partly draws on ISCLOAM lio_sam SC-LEGO-LOAM
# Author of  EPSC-LOAM : QZ Wang  
# Email wangqingzhi27@outlook.com

lis_slam:

  PROJECT_PATH: "/home/wqz/AWorkSpace/LIS-SLAM/src/lis-slam/"
  MODEL_PATH: "/home/wqz/AWorkSpace/darknet53"

  # Export settings
  savePCD: false                                       
  savePCDDirectory: "/AWorkSpace/TestData/map/lis/noLoop/kitti_00/"     # in your home folder, starts and ends with "/".

  saveTrajectory: false
  RESULT_PATH: "/home/wqz/AWorkSpace/TestData/0327/CQU/LIS-SLAM-1-3.txt"

  useImu: false       # If using IMU data, set to "true" 
  useGPSVel : true

  loopClosureEnableFlag: true
  UsingISCFlag : false
  UsingSCFlag : false
  UsingPoseFlag : false
  UsingSEPSCFlag : false
  UsingEPSCFlag : true
  UsingSSCFlag : false
  UsingFEPSCFlag : false


  
  useOdometryPitchPrediction: true              # 在结构性较强时可使用前端里程计pitch预测值

  pointCloudTopic: "rslidar_points"           # Point cloud data
  imuTopic: "imu/data"                        # IMU data
  odomTopic: "odometry"                       # IMU pre-preintegration odometry, same frequency as IMU
  gpsTopic: "odometry/gpsz"                   # GPS odometry topic from navsat, see module_navsat.launch file
  lidarIntensity: "intensity"


  # Frames
  lidarFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"

  # Sensor Settings
  N_SCAN:  16                               # number of lidar channel (i.e., 16, 32, 64, 128)
  Horizon_SCAN: 1800                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048)


  downsampleRate: 1                          # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1 
  lidarMinRange: 0.0                         # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 70.0                        # 70.0      # default: 1000.0, maximum lidar range to be used

  # GPS Settings
  useImuHeadingInitialization: false          # if using GPS data, set to "true"
  useGpsElevation: false                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 25.0                      # m^2, threshold for using GPS data

  # IMU Settings
  imuAccNoise: 3.7686306102624571e-02
  imuGyrNoise: 2.3417543020438883e-03
  imuAccBiasN: 1.1416642385952368e-03
  imuGyrBiasN: 1.4428407712885209e-05
  imuGravity: 9.80511
  imuRPYWeight: 0.01

  # Extrinsics (IMU -> lidar)
  extrinsicTrans: [0.27255, -0.00053,0.17954]
  extrinsicRot: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]
  extrinsicRPY: [1,  0, 0,
                 0, 1, 0,
                  0, 0, 1]
  # extrinsicRot: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]
  # extrinsicRPY: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]
  
  # LOAM feature threshold
  edgeThreshold: 1.0
  surfThreshold: 0.1
  edgeFeatureMinValidNum: -1 #10
  surfFeatureMinValidNum: 100

  # robot motion constraint (in case you are using a 2D robot)
  z_tollerance: 1000                            # meters
  rotation_tollerance: 1000                     # radians

  # CPU Params
  numberOfCores: 2                              # number of cores for mapping optimization
  mappingProcessInterval: 0.15                  # seconds, regulate mapping frequency

  # voxel filter paprams
  odometrySurfLeafSize: 0.4                     # default: 0.4 - outdoor, 0.2 - indoor
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor          

  # Surrounding map
  surroundingKeyframeDensity: 2.0                      # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius:  20.0               # 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  #新增 make Submap
  keyFrameMiniDistance: 0.6  # 0.2-indoor    1.5 kitti 
  keyFrameMiniYaw : 0.3      # 0.2-indoor    0.5-kitti

  #新增 submap voxel filter 
  subMapCornerLeafSize: 0.2                    
  subMapSurfLeafSize: 0.4   
  subMapLeafSize: 0.5   

  subMapYawMax: 0.9          # 0.6-indoor    0.9outdoor
  subMapFramesSize: 20       # 10
  subMapMaxTime: 3.0         # 3.0
  subMapTraMax: 10.0

  subMapOptmizationFirstSize: 2    #5

  # subMapOptmizationWeights: 0.3  #0.3
  odometerAndOptimizedDistanceDifference: 1.02 # meters
  odometerAndOptimizedAngleDifference: 0.01  #  radians

  # Loop closure
  loopClosureFrequency: 2                    # 0.5Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                  # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0            # 15.0       # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0          # 30.0     # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                 # 25        # number of hostory key frames will be fused into a submap for loop closure
  
  historyKeyframeFitnessScore: 0.5            # 0.4  # 0.7  # icp threshold, the smaller the better alignment


  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density





# Navsat (convert GPS coordinates to Cartesian)
navsat:
  frequency: 50
  wait_for_datum: false
  delay: 0.0
  magnetic_declination_radians: 0
  yaw_offset: 0
  zero_altitude: true
  broadcast_utm_transform: false
  broadcast_utm_transform_as_parent_frame: false
  publish_filtered_gps: false

# EKF for Navsat
ekf_gps:
  publish_tf: false
  map_frame: map
  odom_frame: odom
  base_link_frame: base_link
  world_frame: odom

  frequency: 50
  two_d_mode: false
  sensor_timeout: 0.01
  # -------------------------------------
  # External IMU:
  # -------------------------------------
  imu0: imu_correct
  # make sure the input is aligned with ROS REP105. "imu_correct" is manually transformed by myself. EKF can also transform the data using tf between your imu and base_link
  imu0_config: [false, false, false,
                true,  true,  true,
                false, false, false,
                false, false, true,
                true,  true,  true]
  imu0_differential: false
  imu0_queue_size: 50 
  imu0_remove_gravitational_acceleration: true
  # -------------------------------------
  # Odometry (From Navsat):
  # -------------------------------------
  odom0: odometry/gps
  odom0_config: [true,  true,  true,
                 false, false, false,
                 false, false, false,
                 false, false, false,
                 false, false, false]
  odom0_differential: false
  odom0_queue_size: 10

  #                            x     y     z     r     p     y   x_dot  y_dot  z_dot  r_dot p_dot y_dot x_ddot y_ddot z_ddot
  process_noise_covariance: [  1.0,  0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    1.0,  0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    10.0, 0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0.03, 0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0.03, 0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0.1,  0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0.25,  0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0.25,  0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0.04,  0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0.01, 0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0.01, 0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0.5,  0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0.01, 0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0.01,   0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0.015]